core/internal/filetransfer/file_transfer_azure.go

```diff
@@ -169,14 +194,116 @@ func setupBlobClient(
 	return client, nil
 }
 
-// Upload uploads a file to the server. This method is not yet implemented, but
-// is required for the FileTransfer interface.
-func (ft *AzureFileTransfer) Upload(task *ReferenceArtifactUploadTask) error {
-	ft.logger.Debug("Azure file transfer: uploading file", "path", task.PathOrPrefix)
+// Upload uploads a file to the server.
+func (ft *AzureFileTransfer) Upload(task *DefaultUploadTask) error {
+	ft.logger.Debug("Azure file transfer: uploading file", "path", task.Path, "url", task.Url)
+
+	// open the file for reading and defer closing it
+	file, err := os.Open(task.Path)
+	if err != nil {
+		return err
+	}
+	defer func(file *os.File) {
+		err := file.Close()
+		if err != nil {
+			ft.logger.CaptureError(
+				fmt.Errorf(
+					"azure file transfer: upload: error closing file %s: %v",
+					task.Path,
+					err,
+				))
+		}
+	}(file)
+
+	requestBody, err := getUploadRequestBody(task, file, ft.fileTransferStats, ft.logger)
+	if err != nil {
+		return err
+	}
+
+	resp, err := ft.uploadBlob(task, requestBody)
+	if err != nil {
+		return err
+	}
+	task.Response = resp
 
 	return nil
 }
 
+// uploadBlob uploads the given request body to a blob at task.Url.
+func (ft *AzureFileTransfer) uploadBlob(task *DefaultUploadTask, requestBody io.Reader) (*http.Response, error) {
+	clientOptions := blockblob.ClientOptions{
+		ClientOptions: azcore.ClientOptions{
+			Retry: policy.RetryOptions{
+				MaxRetries: 0,
+			},
+		},
+	}
+	blockBlobClient := ft.blockBlobClient
+	if blockBlobClient == nil {
+		client, err := blockblob.NewClientWithNoCredential(task.Url, &clientOptions)
+		if err != nil {
+			return nil, err
+		}
+		blockBlobClient = client
+	}
+
+	uploadOptions := blockblob.UploadStreamOptions{
+		Concurrency: 4,
+		BlockSize:   4 * 1024,
+		HTTPHeaders: &blob.HTTPHeaders{},
+	}
+
+	for _, header := range task.Headers {
+		parts := strings.SplitN(header, ":", 2)
+		if len(parts) != 2 {
+			ft.logger.Error("file transfer: upload: invalid header", "header", header)
+			continue
+		}
+		switch parts[0] {
+		case "Content-MD5":
+			md5, err := base64.StdEncoding.DecodeString(parts[1])
+			if err != nil {
+				return nil, err
+			}
+			uploadOptions.HTTPHeaders.BlobContentMD5 = md5
+		case "Content-Type":
+			uploadOptions.HTTPHeaders.BlobContentType = &parts[1]
+		}
+	}
+
+	resp, err := blockBlobClient.UploadStream(context.Background(), requestBody, &uploadOptions)
+	if err != nil {
+		return nil, err
+	}
+	return &http.Response{
+		StatusCode: 200,
+		Status:     "OK",
+		Header:     getHeadersFromResponse(resp),
+	}, nil
```

Comment:
why do you overwrite the response the you get back from azure?

Reply:
i.e. is it always a 200 status?

Reply:
the response field on the task object expects an http.Response, but the azure response doesn't actually tell us what the status is. it looks like it only ever errors out if the request isn't successfully made, so I just set it as 200 and added all of the headers from the response in case we want to use it later